cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc

```diff
@@ -705,21 +721,75 @@ year_month_day GetFlooredYmd(int64_t arg, int multiple, Localizer localizer_) {
     } else {
       total_months = (total_months - multiple + 1) / multiple * multiple;
     }
-    return year_month_day(year{1970} / jan / 0) + months{total_months};
+    return year{1970} / jan / 1 + months{total_months};
   }
 }
 
 template <typename Duration, typename Unit, typename Localizer>
-const Duration FloorTimePoint(const int64_t arg, const int64_t multiple,
+const Duration FloorTimePoint(const int64_t arg, const RoundTemporalOptions options,
                               Localizer localizer_, Status* st) {
   const auto t = localizer_.template ConvertTimePoint<Duration>(arg);
-  const Unit d = floor<Unit>(t).time_since_epoch();
 
-  if (multiple == 1) {
+  if (options.multiple == 1) {
+    const Unit d = floor<Unit>(t).time_since_epoch();
     return localizer_.template ConvertLocalToSys<Duration>(duration_cast<Duration>(d),
                                                            st);
+  } else if (options.calendar_based_origin) {
+    const Unit unit = Unit{options.multiple};
+
+    switch (options.unit) {
+      case compute::CalendarUnit::DAY: {
+        const auto origin =
+            localizer_.ConvertDays(year_month_day(floor<days>(t)).year() /
+                                   year_month_day(floor<days>(t)).month() / 1);
+        const auto m = (floor<days>(t) - origin) / unit * unit + origin;
+        return localizer_.template ConvertLocalToSys<Duration>(
+            duration_cast<Duration>(m.time_since_epoch()), st);
+      }
+      case compute::CalendarUnit::HOUR: {
+        const auto origin = localizer_.ConvertDays(year_month_day(floor<days>(t)));
+        const auto m = (t - origin) / unit * unit + origin;
+        return localizer_.template ConvertLocalToSys<Duration>(
+            duration_cast<Duration>(m.time_since_epoch()), st);
```

Comment:
This snippet seems repeated in every `case` clause, perhaps move it out of the `switch` entirely?

Reply:
Refactored.